### PR TITLE
expression: refactor grouping function implementation

### PR DESCRIPTION
**Pull request by AilinKid**
### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tiflash/issues/7590

Problem Summary:
As described above, https://github.com/pingcap/tipb/pull/308 tipb PR

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
expression: refactor grouping function implementation
```

### CODE REVIEW

File: dbms/src/Functions/FunctionsGrouping.h [hunk 1]
```diff
@@ -146,31 +158,58 @@ class FunctionGrouping : public IFunctionBase
 
     ResultType groupingImplModeAndBit(UInt64 grouping_id) const
     {
-        return (grouping_id & meta_grouping_id) != 0;
+        UInt64 final = 0;
```
Comment (xzhangxian1008) [inline dbms/src/Functions/FunctionsGrouping.h:161]: Do not use keyword as variable name.

File: dbms/src/Functions/FunctionsGrouping.h [hunk 2]
```diff
@@ -191,6 +230,8 @@ class FunctionBuilderGrouping : public IFunctionBuilder
 
     String getName() const override { return name; }
     bool useDefaultImplementationForNulls() const override { return true; }
+    // at frontend, grouping function can receive maximum number of parameters as 64.
+    // at backend, grouping function has been rewritten as receive only gid with meta.
     size_t getNumberOfArguments() const override { return 1; }
```
Comment (xzhangxian1008) [inline dbms/src/Functions/FunctionsGrouping.h:235]: As the number of input parameter is variable, the return number should be changed

File: dbms/src/Functions/FunctionsGrouping.h [hunk 3]
```diff
@@ -146,31 +158,58 @@ class FunctionGrouping : public IFunctionBase
...
 
     // In grouping function, the number of rolled up columns usually very small,
     // so it's appropriate to use std::set as it is faster than unordered_set in
     // small amount of elements.
-    std::set<UInt64> meta_grouping_marks = {};
+    // one more dimension for multi grouping function args like: grouping(x,y,z...)
+    std::vector<std::set<UInt64>> meta_grouping_marks = {};
 };
 
 class FunctionBuilderGrouping : public IFunctionBuilder
```
Comment (xzhangxian1008) [inline dbms/src/Functions/FunctionsGrouping.h:215]: In `getReturnTypeImpl` function, argument number should be re-considered.

File: dbms/src/Functions/FunctionsGrouping.h [hunk 4]
```diff
@@ -41,7 +41,7 @@ extern const int TOO_MANY_ARGUMENTS_FOR_FUNCTION;
     return (num & (num - 1)) == 0;
 }
 
-using ResultType = UInt8;
+using ResultType = UInt64;
 
 class FunctionGrouping : public IFunctionBase
```
Comment (xzhangxian1008) [inline dbms/src/Functions/FunctionsGrouping.h:46]: In `processGroupingIDs` function, it should receive a vector parameter for grouping ids.

File: dbms/src/Functions/FunctionsGrouping.h [hunk 5]
```diff
@@ -146,31 +158,58 @@ class FunctionGrouping : public IFunctionBase
...
+        {
+            final <<= 1;
+            if ((grouping_id & one_grouping_id) > 0)
+                // col is not needed, meaning being filled null and grouped. = 1
+                final += 1;
+        }
+        return final;
     }
 
     ResultType groupingImplModeNumericCmp(UInt64 grouping_id) const
```
Comment (xzhangxian1008) [inline dbms/src/Functions/FunctionsGrouping.h:172]: ditto

File: dbms/src/Functions/FunctionsGrouping.h [hunk 6]
```diff
@@ -146,31 +158,58 @@ class FunctionGrouping : public IFunctionBase
...
+        {
+            final <<= 1;
+            if (grouping_id > one_grouping_id)
+                // col is not needed, meaning being filled null and grouped. = 1
+                final += 1;
+        }
+        return final;
     }
 
     ResultType groupingImplModeNumericSet(UInt64 grouping_id) const
```
Comment (xzhangxian1008) [inline dbms/src/Functions/FunctionsGrouping.h:185]: ditto

File: dbms/src/Functions/FunctionsGrouping.h [hunk 7]
```diff
@@ -146,31 +158,58 @@ class FunctionGrouping : public IFunctionBase
 
     ResultType groupingImplModeAndBit(UInt64 grouping_id) const
```
Comment (xzhangxian1008) [inline dbms/src/Functions/FunctionsGrouping.h:159]: ditto

File: dbms/src/Functions/tests/gtest_grouping.cpp
Comment (xzhangxian1008) [inline dbms/src/Functions/tests/gtest_grouping.cpp:1]: I think we need to add more cases when we have multi parameters

File: dbms/src/Functions/FunctionsGrouping.h (hunk 2 above)
Comment (AilinKid) [reply dbms/src/Functions/FunctionsGrouping.h:235]: It's ok here, the pb function that tiflash received only has one arg (gid) with meta

File: dbms/src/Functions/FunctionsGrouping.h (hunk 3 above)
Comment (AilinKid) [reply dbms/src/Functions/FunctionsGrouping.h:215]: as said in the comment above, the pb function that tiflash received only has one arg (gid) with metas, so `getReturnTypeImpl` seems good

File: dbms/src/Functions/FunctionsGrouping.h (hunk 4 above)
Comment (AilinKid) [reply dbms/src/Functions/FunctionsGrouping.h:46]: not really, as commented above,
grouping(a,b,c,xxxx, n) after tidb's rewriting, it changed as grouping(gid) [meta1, meta2, ... metan], it grouping id column is still exactly one, while we should do multi computation between gid and metaX, then get the final result.

File: dbms/src/Functions/tests/gtest_grouping.cpp
Comment (AilinKid) [reply dbms/src/Functions/tests/gtest_grouping.cpp:1]: sure

**Review by xzhangxian1008 — APPROVED** [review]
LGTM

**Review by guo-shaoge — APPROVED** [review]

**Comment by AilinKid** [comment]
Issue Number: close #7590

File: dbms/src/Functions/FunctionsGrouping.h [hunk 8]
```diff
@@ -146,31 +158,58 @@ class FunctionGrouping : public IFunctionBase
...
     {
-        return (grouping_id & meta_grouping_id) != 0;
+        UInt64 res = 0;
+        for (auto one_grouping_id : meta_grouping_ids)
+        {
+            res <<= 1;
+            if ((grouping_id & one_grouping_id) <= 0)
```
Comment (windtalker) [inline dbms/src/Functions/FunctionsGrouping.h:165]: Why `(grouping_id & one_grouping_id) <= 0`, I think `(grouping_id & one_grouping_id) < 0` is always false, so the condition could be `(grouping_id & one_grouping_id) == 0`?

**Review by windtalker — APPROVED** [review]
Others LGTM

File: dbms/src/Functions/FunctionsGrouping.h (hunk 8 above)
Comment (AilinKid) [reply dbms/src/Functions/FunctionsGrouping.h:165]: yes, I make `!((grouping_id & one_grouping_id) > 0)`, this is the lint change

**Pull request merged**